server/render/index.js

```diff
@@ -130,7 +130,15 @@ export function serverRender( req, res ) {
 		cacheKey = false;
 
 	if ( isSectionIsomorphic( context.store.getState() ) && ! context.user ) {
-		cacheKey = getCacheKey( context );
+		if ( isEmpty( context.query ) ) {
+			cacheKey = context.pathname;
+		} else {
+			// If we have query args, make sure we only cache if they're whitelisted
+			const cachedQueryParams = pick( context.query, context.cacheQueryKeys );
+			if ( ! isEmpty( cachedQueryParams ) ) {
+				cacheKey = getNormalizedPath( context.pathname, cachedQueryParams );
+			}
+		}
```

Comment:
I spent a long time studying this, thinking that we didn't need the first `isEmpty( context.query )` check because of the check inside `getNormalizedPath`. I see now that it is necessary for the whitelisting.

I suppose it could be condensed to:
```
// If we have query args, make sure we only cache if they're whitelisted
const cachedQueryParams = pick( context.query, context.cacheQueryKeys );
if (  isEmpty( context.query || ! isEmpty( cachedQueryParams ) ) {
	cacheKey = getNormalizedPath( context.pathname, cachedQueryParams );
}
```
But that is not any less confusing, so it is fine as you have it I would say.

Reply:
Yeah, I've gone thru the exact same pondering :laughing:

As described at https://github.com/Automattic/wp-calypso/pull/22593#issuecomment-381626579, I think the check needs to a bit different from `! isEmpty( cachedQueryParams )`, so the whole thing might change a little anyway (might become more elegant, since I'm thinking of essentially comparing `isEqual( sort( Object.keys( context.query ) ), sort( cachedQueryParams ) )`, which includes the empty case (if I also provide a default value of `[]` for `cachedQueryParams`)